### PR TITLE
nn: add missing memory_format overload to Module.to()

### LINKED ISSUE #184642: `Module.to()` has no overload for `.to(memory_format=torch.channels_last)`
### 🐛 Describe the bug

`Module.to()`'s documents and supports being called with only a `memory_format` argument:

```
This can be called as
[...]
.. function:: to(memory_format=torch.channels_last)
```

This works at runtime, but the method's type signature has no matching overload. This causes issues with type checkers like Astral's `ty`.

Minimal repro:

```python
import torch

module = torch.nn.Module()
module.to(memory_format=torch.channels_last)  # ty(no-matching-overload)
```

An overload like the following should exist:
```python
@overload
def to(self, memory_format: torch.memory_format) -> Self: ...
```

<details>
<summary>Versions</summary>
Collecting environment information...
PyTorch version: 2.11.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.12.3 (main, Mar 23 2026, 19:04:32) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-110-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 13.0.88
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 4090
GPU 1: NVIDIA GeForce RTX 4090

Nvidia driver version: 595.58.03
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  32
On-line CPU(s) list:                     0-31
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Core(TM) i9-14900K
CPU family:                              6
Model:                                   183
Thread(s) per core:                      2
Core(s) per socket:                      24
Socket(s):                               1
Stepping:                                1
CPU(s) scaling MHz:                      21%
CPU max MHz:                             6000.0000
CPU min MHz:                             800.0000
BogoMIPS:                                6374.40
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities ibpb_exit_to_user
Virtualization:                          VT-x
L1d cache:                               896 KiB (24 instances)
L1i cache:                               1.3 MiB (24 instances)
L2 cache:                                32 MiB (12 instances)
L3 cache:                                36 MiB (1 instance)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-31
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Mitigation; Clear Register File
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect
</details>

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @lolpack @maggiemoss @ndmitchell @kinto0 @samwgoldman @jamesr66a

### PR DESCRIPTION
**Pull request by Priyanshu-byte-coder**
## Summary

`Module.to(memory_format=torch.channels_last)` is documented and works at runtime, but there was no matching `@overload` for type checkers. This causes `no-matching-overload` errors in tools like `ty` (Astral).

Add the missing overload and import `memory_format` from `torch`.

Fixes #184642.

### CODE REVIEW

**Comment by Priyanshu-byte-coder** [comment]
@pytorch-dev-infra could you approve CI and add release notes label? This adds a missing type overload — no behavior change.

**Comment by Priyanshu-byte-coder** [comment]
@huydhn @frgossen could you take a look? Adds a missing `@overload` for `Module.to(memory_format=...)` — type checkers can't currently infer the return type for this call form.

**Comment by Priyanshu-byte-coder** [comment]
Still relevant. This adds the missing @overload for Module.to(memory_format=...); without it type checkers can't infer the return type for that call form. Typing-only change, no runtime impact. @huydhn @frgossen could someone take a look?

**Review by albanD — APPROVED** [review]
Not sure how much this changes, but thanks for taking the time to send a PR.

**Comment by albanD** [comment]
## mergedog handoff

All CI is green (or skipped). Ready for human review and `@pytorchbot merge`.


Current PR head: `93beaacb3dc632c3eeb0df2cb6362ad5c3e8758e`.
No mergedog-authored commits were pushed during this run.

No LLM was invoked during this run.
